common/models/transducer/rna_pure_tf.py

```diff
@@ -0,0 +1,641 @@
+      sel_emit = tf.stack([u_ranged_shifted, labels_emit], axis=-1)  # (B, n) | (B,)
+      sel_same = tf.stack([u_ranged, labels_same], axis=-1)
+
+      argmax_idx_tiled = tf.tile(argmax_idx[..., tf.newaxis], [1, 1, 2])  # (B, n, 2)
+      sel = tf.where(tf.equal(argmax_idx_tiled, 0),
+                     sel_blank,  # blank
+                     tf.where(tf.equal(argmax_idx_tiled, 1),
+                              sel_emit,  # emit
+                              sel_same))  # same
+      # u_ranged, u_ranged_shifted)  # (B, U|n)
+      # we need to pad so we can stack the TA later on (instead of triangular shape)
+      sel_padded = tf.pad(sel, [[0, 0], [0, max_target - max_len], [0, 0]])
+      bt_ta = bt_ta.write(n - 1, sel_padded)
+    else:
+      bt_ta = None
+
+    red_op = tf.stack(reduction_args, axis=0)  # (2, B, N)
+    red_op = py_print_iteration_info("red-op", red_op, n, debug=debug)
+    new_column = tf.math.reduce_logsumexp(red_op, axis=0)  # (B, N)
+
+    new_column = new_column[:, :n]
+    new_column = py_print_iteration_info("new_column", new_column, n, debug=debug)
+    ret_args = [n + 1, alpha_ta.write(n, new_column)]
+    return ret_args + [bt_ta] if with_alignment else ret_args
+
+  n = tf.constant(2)
+  initial_vars = [n, alpha_ta]
+  if with_alignment:
+    initial_vars += [bt_ta]
+  final_loop_vars = tf.while_loop(cond, body_forward,
+                                  initial_vars,
+                                  parallel_iterations=1,  # iterative computation
+                                  name="rna_loss")
+  if with_alignment:
+    final_n, alpha_out_ta, bt_ta = final_loop_vars
+    bt_mat = tf.transpose(bt_ta.stack(), [1, 0, 2, 3])  # (T, B, U, 2) -> (B, T, U, 2)
+    alignments = compute_alignment_tf(bt_mat, input_lengths, label_lengths + 1)
+  else:
+    final_n, alpha_out_ta = final_loop_vars
+  # p(y|x) = alpha(T,U) * blank(T,U)  (--> in log-space)
+  # ll_tf = final_alpha[n_time-1, n_target-1]
+
+  # (B,): batch index -> column index
+  col_idxs = input_lengths + 1   # (B,)
+
+  # (B,): batch index -> index within column
+  within_col_idx = label_lengths
+  within_col_idx = tf.where(tf.less_equal(label_lengths, input_lengths),
+                            within_col_idx,  # everything ok, T>U
+                            tf.ones_like(within_col_idx) * -1)  # U > T, not possible in RNA
+
+  res_ta = tf.TensorArray(
+    dtype=tf.float32,
+    clear_after_read=True,
+    size=n_batch,
+    dynamic_size=False,
+    infer_shape=False,
+    element_shape=(),
+    name="alpha_columns",
+  )
+  tf_neg_inf = tf.constant(NEG_INF)
+
+  def ta_read_body(i, res_loop_ta):
+    """Reads from the alpha-columns TensorArray. We need this because of the inconsistent shapes in the TA."""
+    ta_item = alpha_out_ta.read(col_idxs[i])[i]
+    elem = tf.cond(tf.equal(within_col_idx[i], -1), lambda: tf_neg_inf, lambda: ta_item[within_col_idx[i]])
```

Comment:
It seems like no backward-forward is used, only forward. Here the element `alpha[T, N]` is taken as the loss.

Reply:
You only need the forward dynamic programming calculation to calculate the loss (log prob). The backward pass is implicitly automatic via automatic differentiation.

Reply:
I read that at Andre`s thesis too. Can't understand from the code where the backward matrix is used/mentioned.
I though't you multiply the forward pass matrix and backward pass matrix over a diagonal and take the sum. But here it is the extreme case where only the forward pass matrix is used and only the element on the top-right is taken(i.e. sum of the  last diagonal, which has length 1)

Reply:
As I said, you do not need the backward pass. It is automatic (via automatic differentiation). It follows from the chain rule.

You only need to calculate the loss (log prob). And for that, you only need the forward pass.

In the last frame, you only take the score from the valid ending state. Which is the element at the top-right.

So all seems actually correct.